controller/src/main/java/io/pravega/controller/store/stream/records/SubscriberSet.java

```diff
@@ -37,37 +37,55 @@
     public static final SubscriberSetSerializer SERIALIZER = new SubscriberSetSerializer();
 
     @Getter
-    private final ImmutableList<String> subscribers;
+    private final ImmutableMap<String, Long> subscribers;
 
     @Builder
-    public SubscriberSet(@NonNull ImmutableList<String> subscribers) {
+    public SubscriberSet(@NonNull ImmutableMap<String, Long> subscribers) {
         this.subscribers = subscribers;
     }
 
     /**
      * This method adds a subscriber in the subscriberSet.
      * @param subscriberSet Subscriber Set.
      * @param subscriber subscriber to be added.
+     * @param generation subscriber generation.
      * @return updated Subscriber Set.
      */
-    public static SubscriberSet add(@NonNull SubscriberSet subscriberSet, @NonNull String subscriber) {
-            ImmutableList.Builder<String> builder = ImmutableList.builder();
-            builder.addAll(subscriberSet.subscribers);
-            builder.add(subscriber);
+    public static SubscriberSet add(@NonNull SubscriberSet subscriberSet, @NonNull String subscriber, long generation) {
+            ImmutableMap.Builder<String, Long> builder = ImmutableMap.builder();
+            builder.putAll(subscriberSet.subscribers);
+            builder.put(subscriber, generation);
             return new SubscriberSet(builder.build());
     }
 
+    /**
+     * This method updates the generation of a subscriber in the subscriberSet.
+     * @param subscriberSet Subscriber Set.
+     * @param subscriber subscriber to be added.
+     * @param generation subscriber generation.
+     * @return updated Subscriber Set.
+     */
+    public static SubscriberSet update(@NonNull SubscriberSet subscriberSet, @NonNull String subscriber, long generation) {
+        ImmutableMap.Builder<String, Long> builder = ImmutableMap.builder();
+        Map<String, Long> otherSubscribers = subscriberSet.getSubscribers().entrySet().stream().filter(e -> !e.getKey().equals(subscriber))
```

Comment:
this can be improved.. its making two passes (filter and putall) over the map. we can get it down to one pass only by adding all the existing entries and for entry which already exists, we add the new subscriber.

Reply:
ok.